Source/Core/Core/IOS/Network/KD/VFF/VFFUtil.cpp

```diff
@@ -0,0 +1,261 @@
+    return RES_ERROR;
+  }
+
+  const size_t size = static_cast<size_t>(count) * SECTOR_SIZE;
+  if (!s_vff->Read(buff, size))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF read failed (offset={}, size={})", offset, size);
+    return RES_ERROR;
+  }
+
+  return RES_OK;
+}
+
+extern "C" DRESULT vff_write(BYTE pdrv, const BYTE* buff, LBA_t sector, UINT count)
+{
+  const u64 offset = static_cast<u64>(sector) * SECTOR_SIZE - 480;
+  if (!s_vff->Seek(offset, IOS::HLE::FS::SeekMode::Set))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF seek failed (offset={})", offset);
+    return RES_ERROR;
+  }
+
+  const size_t size = static_cast<size_t>(count) * SECTOR_SIZE;
+  const auto res = s_vff->Write(buff, size);
+  if (!res)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF write failed (offset={}, size={})", offset, size);
+    return RES_ERROR;
+  }
+
+  return RES_OK;
+}
+
+extern "C" DRESULT vff_ioctl(BYTE pdrv, BYTE cmd, void* buff)
+{
+  switch (cmd)
+  {
+  case CTRL_SYNC:
+    return RES_OK;
+  case GET_SECTOR_COUNT:
+    *reinterpret_cast<LBA_t*>(buff) = s_vff->GetStatus()->size / SECTOR_SIZE;
+    return RES_OK;
+  default:
+    WARN_LOG_FMT(IOS_WC24, "Unexpected FAT ioctl {}", cmd);
+    return RES_OK;
+  }
+}
+
+extern "C" DWORD vget_fattime(void)
+{
+  const std::time_t time = std::time(nullptr);
+  std::tm tm;
+#ifdef _WIN32
+  localtime_s(&tm, &time);
+#else
+  localtime_r(&time, &tm);
+#endif
+
+  DWORD fattime = 0;
+  fattime |= (tm.tm_year - 80) << 25;
+  fattime |= (tm.tm_mon + 1) << 21;
+  fattime |= tm.tm_mday << 16;
+  fattime |= tm.tm_hour << 11;
+  fattime |= tm.tm_min << 5;
+  fattime |= std::min(tm.tm_sec, 59) >> 1;
+  return fattime;
+}
+
+namespace IOS::HLE::NWC24
+{
+
+static ErrorCode WriteFile(const std::string& filename, const std::vector<u8>& tmp_buffer)
+{
+  FIL dst;
+  const auto open_error_code = f_open(&dst, filename.c_str(), FA_CREATE_ALWAYS | FA_WRITE);
+  if (open_error_code != FR_OK)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to open file {} in VFF", filename);
+    return WC24_ERR_FILE_OPEN;
+  }
+
+  u32 written_size;
+  const auto write_error_code = f_write(&dst, tmp_buffer.data(), tmp_buffer.size(), &written_size);
+  if (write_error_code != FR_OK)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to write file {} to VFF {}", filename, write_error_code);
+    return WC24_ERR_FILE_WRITE;
+  }
+
+  const auto close_error_code = f_close(&dst);
+  if (close_error_code != FR_OK)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to close file {} in VFF", filename);
+    return WC24_ERR_FILE_CLOSE;
+  }
+
+  return WC24_OK;
+}
+
+ErrorCode OpenVFF(const std::string& path, const std::string& filename,
+                  const std::shared_ptr<FS::FileSystem>& fs, const std::vector<u8>& data)
+{
+  std::lock_guard lk(s_fatfs_mutex);
+
+  auto file = fs->OpenFile(PID_KD, PID_KD, path, FS::Mode::ReadWrite);
+  if (!file)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to open VFF at: {}", path);
+    return WC24_ERR_NOT_FOUND;
+  }
+
+  s_vff = file.operator->();
```

Comment:
This feels a little dirty (calling an operator by name). But the question is: Does `s_vff = &*file;` look any clearer? IMO: no. (Plus I'm not sure if that even works here.)

I don't suppose we can make this any more clear and obvious to someone looking at the code later, can we?

Reply:
What would you suggest I make s_vff?